ntoskrnl/mm/ARM3/pagfault.c

```diff
@@ -1227,16 +1227,26 @@ MiResolveProtoPteFault(IN BOOLEAN StoreInstruction,
         /* Perform the copy */
         MiCopyPfn(PageFrameIndex, ProtoPageFrameIndex);
 
+        /* Temporarily release lock */
+        MiReleasePfnLock(OldIrql);
+
         /* This will drop everything MiResolveProtoPteFault referenced */
         MiDeletePte(PointerPte, Address, Process, PointerProtoPte);
 
-        /* Because now we use this */
-        Pfn1 = MI_PFN_ELEMENT(PageFrameIndex);
-        MiInitializePfn(PageFrameIndex, PointerPte, TRUE);
-
         /* Fix the protection */
         Protection &= ~MM_WRITECOPY;
         Protection |= MM_READWRITE;
+
+        /* Restore our PTE to a sensible value for our PFN */
+        PointerPte->u.Long = MmProtectToPteMask[Protection];
```

Comment:
Use `` MI_UPDATE_VALID_PTE``

Reply:
I don't think that we want to keep the old mapping in Pfn->OriginalPte.

Reply:
On second thought, we don't want the hardware bits here, but the soft PTE with the proper protection. Corrected.

Reply:
I still think this shouldn't just be updated like this. There is a reason for using the MI_*_PTE functions. Simply writing to PTEs without any checks should not be done unless there is a real reason to do so, which should be provided in a comment.